### ISSUE TITLE
resolvers and load-server-state-from-file do not work as expected

### ISSUE THREAD

**Issue by gabrieltz** [haproxy/haproxy#1387]
### Detailed Description of the Problem

While upgrading multiple instances to 2.0.25 today I noticed that when load-server-state-from-file is used along with server-template and resolver after restarting haproxy the server-state that was loaded may be as old as the last systemctl reload command and some stale server records are loaded and marked as down . Since we're using resolvers, the server entries should eventually be updated but this never happened.

### Expected Behavior

The expected behaviour, even if the server state file is out of date, should be to eventually update the servers in the backend since it receives valid dns responses . While i can capture dns traffic about specific SRV entries between haproxy and consul , the server entries are not updated and the old ones are marked as DOWN and are never updated from DNS.

### Steps to Reproduce the Behavior

1. save server state 
2. make changes to backends dynamicly from resolvers/dnsmasq so that backend servers get shuffled and moved.
3. restart haproxy with state server-state fle.

### Do you have any idea what may have caused this?

the systemd unit file saves the server state upon systemd reload with socat , but the restart does not .
So the server state file may be very old and servers' IPs and random ports may have changed.

```
[Unit]
Description=HAProxy Load Balancer
After=network.target

[Service]
Environment="CONFIG=/etc/haproxy/haproxy.cfg" "PIDFILE=/run/haproxy.pid"
ExecStartPre=/usr/sbin/haproxy -f $CONFIG -c -q
ExecStart=/usr/sbin/haproxy -Ws -f $CONFIG -p $PIDFILE
ExecReload=/usr/sbin/haproxy -f $CONFIG -c -q
ExecReload=/bin/bash -c "/usr/bin/socat /run/haproxy/admin.sock - <<< 'show servers state' > /etc/haproxy/state/current"
ExecReload=/bin/kill -USR2 $MAINPID
KillMode=mixed
Restart=always
SuccessExitStatus=143
Type=notify

[Install]
WantedBy=multi-user.target
```

### Do you have an idea how to solve the issue?

2 ideas: the third is software fixes which i sadly can not do

1. issueing reload before restart ensures the server state file is current .
2. load-server-state-from-file none in affected backends and reload again.
3.  Maybe haproxy should only use the server state file 
as a starting point to fill the server-templates and then do it's thing using the resolvers' returned values to replace dead backend servers.

### What is your configuration?

```haproxy
global
	chroot /var/lib/haproxy
	stats socket /run/haproxy/admin.sock mode 660 level admin
        stats socket ipv4@10.1.1.99:9999 level admin
	stats timeout 30s
	maxconn 10000
	log /dev/log/ local3 notice
	server-state-file /etc/haproxy/state/current
	user haproxy
	group haproxy
	daemon
	nbproc 1

	ca-base /etc/ssl/certs
	crt-base /etc/ssl/private

	ssl-default-bind-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:ECDH+3DES:DH+3DES:RSA+AESGCM:RSA+AES:RSA+3DES:!aNULL:!MD5:!DSS
	ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11
	tune.ssl.default-dh-param 1024
	tune.ssl.maxrecord 1350

defaults
	mode	http
	option	dontlognull
        timeout connect 5s
        timeout client  60s
        timeout server  60s
	errorfile 400 /etc/haproxy/errors/400.http
	errorfile 403 /etc/haproxy/errors/403.http
	errorfile 408 /etc/haproxy/errors/408.http
	errorfile 500 /etc/haproxy/errors/500.http
	errorfile 502 /etc/haproxy/errors/502.http
	errorfile 503 /etc/haproxy/errors/503.http
	errorfile 504 /etc/haproxy/errors/504.http
	log /dev/log local3 notice
	load-server-state-from-file global
	backlog 131072

resolvers consul
        nameserver consulA     10.1.1.1:8600
        nameserver consulB     10.1.1.2:8600
        nameserver consulC     10.1.1.3:8600
        accepted_payload_size 8192
        resolve_retries       3
        timeout resolve       1s
        timeout retry         2s

frontend cluster-ssl
        mode http
        option  httplog
        option  forwardfor

        bind 127.0.0.1:80 
        use_backend %[str(active),map(/etc/haproxy/livecluster.map,cluster-blue)] if { hdr_dom(host) mydomain }


backend cluster-green
        mode http
	fullconn 80
	load-server-state-from-file none
        http-response add-header  X-Color green
        server-template green-cluster 4  _cluster-green._tcp.service.consul maxconn 20 weight 50 resolvers consul check observe layer4 error-limit 4 on-error mark-down

backend cluster-blue
        mode http
	fullconn 80
	load-server-state-from-file none
        http-response add-header X-Color blue
        server-template blue-cluster 4 _cluster-blue._tcp.service.consul maxconn 20 weight 50 resolvers consul check observe layer4 error-limit 4 on-error mark-down
```
systemd unit already pasted above.

### Output of `haproxy -vv`

```plain
HA-Proxy version 2.0.25-6986403 2021/09/07 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = native
  CC      = gcc
  CFLAGS  = -O2 -march=native -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wno-stringop-overflow -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_THREAD=1 USE_OPENSSL=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL -LUA +FUTEX +ACCEPT4 -CLOSEFROM -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=12).
Built with OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
Running on OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.31 2018-02-12
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace
```

### Additional Information

I only have one server running  2.4.4-acb1d0b and the issue did not manifest there having a server-state file that is at least a week old and plenty of topology changes occured since then . 

Back to haproxy 2.0.25
What may help you is that for a specific backend with 4 servers ( from server template ) that consul keeps track of
is that i have 2 running and 2 dead ones and this should be 3 running 1 down .
The active ones have srv_op_state = 2  , and one having srv_admin_state 0 and the other 64 
The Down ones have srv_op_state = 0 and srv_admin_state =  0

so only one was set from resolvers it's probably we use 3 out 4 instances so the state for the fourth was empty.|
the rest were loaded from the state file and by luck one of them was still active.

**Comment by gabrieltz** [comment]
While upgrading our standby lan balancer this morning I saved the server-state-file and i am attaching it, along with the `show servers state ` output AFTER the restart and some dns packets between haproxy  and consul which shows that proper SRV records are returned . While 2 out of 3 backend servers match by hostname, the ports are different so haproxy has all 3 backend servers marked as down.  Due to it's nature, the standby haproxy is rarely reloaded so the state file was having very old server entries.

![Screenshot from 2021-09-10 09-55-21](https://user-images.githubusercontent.com/13848662/132813263-50d113ed-ccd7-47e8-8538-4700f2f0d5e4.png)

[servers-state-file-before_restart.txt](https://github.com/haproxy/haproxy/files/7141865/servers-state-file-before_restart.txt)
[show-servers-state-after_restart.txt](https://github.com/haproxy/haproxy/files/7141866/show-servers-state-after_restart.txt)
[syslog.txt](https://github.com/haproxy/haproxy/files/7141893/syslog.txt)


I hope this live example helps you

**Comment by capflam** [comment]
Thanks for report, I'll take a look.

**Comment by capflam** [comment]
After some investigations, I found it is in fact a limitation of the 2.0. The bug was fixed as far as 2.2 but this part is too different on the 2.0 to backport the fixes. The risk to break something else is really to high. This part is a real mess. We tried several times to fix bugs and each time this introduced other bugs.

So, you have now 3 solutions :
* Upgrade to 2.2 or higher
* Be sure to update the state file before any restart to have up to date info (and be sure to not change the DNS records in the mean time).
* If it is not too often and if it is manageable, you can choose to modify it by hand (or via a script) before any restart.

The first one is probably the best. 2.2 is now pretty stable, at least as much as the 2.0.

**Comment by gabrieltz** [comment]
Both 2.2 and 2.4 require me to rewrite most health checks , that's why i was avoiding the upgrade.  There is no alternative now but to upgrade :) 

thank you for your assistance and keep up the good work
